llvm/test/CodeGen/AArch64/wide-scalar-shift-by-byte-multiple-legalization.ll

```diff
@@ -238,18 +270,34 @@ define void @ashr_32bytes(ptr %src.ptr, ptr %byteOff.ptr, ptr %dst) nounwind {
 ; ALL-NEXT:    ldp x9, x8, [x0, #16]
 ; ALL-NEXT:    ldr x10, [x1]
 ; ALL-NEXT:    ldr q0, [x0]
-; ALL-NEXT:    and x10, x10, #0x1f
+; ALL-NEXT:    and x11, x10, #0x18
 ; ALL-NEXT:    stp x9, x8, [sp, #16]
 ; ALL-NEXT:    asr x8, x8, #63
 ; ALL-NEXT:    mov x9, sp
 ; ALL-NEXT:    str q0, [sp]
+; ALL-NEXT:    add x9, x9, x11
 ; ALL-NEXT:    stp x8, x8, [sp, #48]
 ; ALL-NEXT:    stp x8, x8, [sp, #32]
-; ALL-NEXT:    add x8, x9, x10
-; ALL-NEXT:    ldp x10, x9, [x8, #16]
-; ALL-NEXT:    ldr q0, [x8]
-; ALL-NEXT:    str q0, [x2]
-; ALL-NEXT:    stp x10, x9, [x2, #16]
+; ALL-NEXT:    lsl x8, x10, #3
```

Comment:
This is a lot worse?

Reply:
In the legacy code the shift unit is 1 byte. If the shift amount is known to be a multiple of 1 byte, one load with an offset is enough to do the shift. In this test case the shift amount is the multiple of 1 byte.

Now the shift unit for aarch64 is 8 byte, and the shift amount is not a multiple of 8 bytes, so we need a load+shift operation.

I think that this could be fixed only if we let unaligned loads in some cases.

Reply:
If this is a lot worse then why not just allow unaligned loads? It could test allowsMisalignedMemoryAccesses to check if they are supported and the IsFast parameter can tell you if they are efficient.

Reply:
The unaligned load is on the unrelated operation that happened to produce the incoming value. There's no natural reason in context to care about unaligned loads

Reply:
I see, it was in a previous version but taken out. So this makes strict-align better, but systems that allow unaligned access can be worse if the shift amount is known to be at least byte aligned.

Reply:
Maybe we could add a logic the calculate the shifting unit based on the support of unaligned memory access. I was thinking something like this:
`
  Align LoadAlign = [&]() {
    EVT LoadVT = VT;
    do {
      LoadVT = TLI.getTypeToTransformTo(*DAG.getContext(), LoadVT);
    } while (!TLI.isTypeLegal(LoadVT));

    Align LoadAlign(LoadVT.getStoreSize());
    for (Align Tmp = LoadAlign;; Tmp = Tmp.previous()) {
      unsigned IsFast = 0;
      const bool AllowsFastMisalignedMemoryAccesses =
          TLI.allowsMisalignedMemoryAccesses(
              LoadVT, DAG.getDataLayout().getAllocaAddrSpace(), Tmp,
              MachineMemOperand::MOLoad, &IsFast) &&
          IsFast;
      if (not AllowsFastMisalignedMemoryAccesses) {
        break;
      }
      LoadAlign = Tmp;
      if (LoadAlign == Align(1))
        break;
    }
    return LoadAlign;
  }();
  const unsigned ShiftUnitInBits = LoadAlign.value() * 8;
  ...
  // And load it! While the load is not legal, legalizing it is obvious.
  SDValue Res =
      DAG.getLoad(VT, dl, Ch, AdjStackPtr,
                  MachinePointerInfo::getUnknownStack(DAG.getMachineFunction()),
                  commonAlignment(StackAlign, LoadAlign.value()));


`
@arsenm I know that you have already approved the PR, but what do you think?

Reply:
It still doesn't make sense to be to be considering unaligned accesses when there is no unaligned access in the source operation, and no need to introduce one. I don't think anything like this belongs in this patch, and it seems more like a missing optimization elsewhere

Reply:
I'm not sure what is meant by "no need to introduce one". Is it not just a better optimization to to be able to load unaligned back from the stack?

Reply:
The original operation has no stack operation. This is creating a new stack slot that can use any alignment, so why would you introduce a new underaligned slot? I still think any adjustment here would be best done in a follow up PR

Reply:
Am I right is saying that we start off with a large i256 shift, where we can work out that it shifts by a multiple of 8? So we store the original value to the stack slot of twice the size, and reload from `stack+shift/8`? And if you use that alignment you don't need to do anything else, but if you require a higher alignment then you might need to do a follow-up shift too?

Reply:
I'm computing `shift_aligned` value from the load alignment. Then I load from `stack+shift/shift_aligned`, which will be an aligned load now. If I can work out that `shift` is a multiple of `shift_aligned`, we are done, otherwise we need a follow-up shift.

Reply:
Yeah OK. I don't see why we would make this optimization worse by requiring aligned loads on machines where there is no penalty for doing so, but it is relatively unlikely to come up in practice so feel free to ignore me and continue on getting this committed.